esp/bindings/http/platform/httpservice.cpp

```diff
@@ -974,3 +875,538 @@ int CEspHttpServer::onGet()
+        return false;
+
+    ISecResourceList* rlist = authmap->getResourceList(authReq.httpPath.str());
+    if(!rlist) //No auth requirement for the httpPath.
+        return false;
+
+    authReq.ctx->setAuthenticationMethod(authMethod);
+    authReq.ctx->setResources(rlist);
+
+    return true;
+}
+
+EspAuthState CEspHttpServer::checkUserAuthPerSession(EspAuthRequest& authReq)
+{
+    ESPLOG(LogMax, "checkUserAuthPerSession");
+
+    unsigned sessionID = readCookie(authReq.authBinding->querySessionIDCookieName());
+    if (strieq(authReq.httpPath.str(), authReq.authBinding->queryLoginURL()))
+    {//This is a request to ask for a login page.
+        if (sessionID != 0)
+            return authSucceeded;//If there is a valid session ID, now allow to send out the login page.
+
+        ESPLOG(LogMin, "Authentication failed: no session ID found for login page.");
+        //Create a new session and redirect to the login page.
+        askUserLogin(authReq.authBinding, createHTTPSession(authReq, "/"));
+        return authSucceeded;
+    }
+
+    if (sessionID > 0)
+        return doSessionAuth(sessionID, authReq, nullptr, nullptr);//Check session based authentication using this session ID.
+
+    if(!readCookie(SESSION_AUTH_COOKIE))
+        return authUnknown;
+
+    //The session has expired.
+    //If userName/password are received (from login page), we still can check session based authentication.
+    const char* userName = (authReq.requestParams) ? authReq.requestParams->queryProp("username") : NULL;
+    const char* password = (authReq.requestParams) ? authReq.requestParams->queryProp("password") : NULL;
+    if (!isEmptyString(userName) && !isEmptyString(password))
+        return doSessionAuth(createHTTPSession(authReq, "/"), authReq, userName, password);
+
+    if(authReq.isSoapPost) //from SOAP Test page
+        sendMessage("Session expired. Please close this page and login again.", "text/html; charset=UTF-8");
+    else //from other page
+        askUserLogin(authReq.authBinding, createHTTPSession(authReq, nullptr));
+    return authFailed;
+}
+
+EspAuthState CEspHttpServer::checkUserAuthPerRequest(EspAuthRequest& authReq)
+{
+    ESPLOG(LogMax, "checkUserAuthPerRequest");
+
+    authReq.authBinding->populateRequest(m_request.get());
+    if (authReq.authBinding->doAuth(authReq.ctx))
+    {//We do pass the authentication per the request
+        // authenticate optional groups. Do we still need?
+        authOptionalGroups(authReq);
+
+        StringBuffer userName, peer;
+        ESPLOG(LogNormal, "Authenticated for %s@%s", authReq.ctx->getUserID(userName).str(), m_request->getPeer(peer).str());
+        return authSucceeded;
+    }
+    if(!authReq.isSoapPost)
+        return authUnknown;
+
+    //If SoapPost, username/password may be in soap:Header which is not in HTTP header.
+    //The doAuth() may check them inside CSoapService::processHeader() later.
+    authReq.ctx->setToBeAuthenticated(true);
```

Comment:
we should really replace this two step WsSecurity check in the future and just go ahead and look into the soap security headers here.  @wangkx Can you open a new JIRA about that?